File: 220/ruby/template_method/encoder.rb
class Encoder
  
  def encode(input)
    input.split(' ').map { |word| encode_word(word) }.join(" ") 
  end
  
end

class ReverseEncoder < Encoder
  
  def encode_word(word)
    word.reverse
  end
  
end

class DoublingEncoder < Encoder
  
  def encode_word(word)
    word * 2
  end
  
end

File: 220/js/template_method/encoder.js
Encoder = function Encoder() {
	this.encode = function(string) {
		return string.split(" ").map(this.encodeWord).join(" ");
	};
};

ReverseEncoder = function ReverseEncoder() {}
ReverseEncoder.prototype = new Encoder();
ReverseEncoder.prototype.encodeWord = function(word) {
	return word.split("").reverse().join("");
};

DoubleEncoder = function DoubleEncoder() {
	this.encodeWord = function(word) {
		return word + word;
	};
};
DoubleEncoder.prototype = new Encoder();